circuits/circom/trees/incrementalQuinTree.circom

```diff
@@ -0,0 +1,249 @@
+        muxes[i].s <== isLeafIndex[i].out;
+        muxes[i].c[0] <== quinSelectors[i].out;
+        muxes[i].c[1] <== leaf;
+
+        out[i] <== muxes[i].out;
+    }
+}
+
+template QuinTreeInclusionProof(levels) {
+    // Each node has 5 leaves
+    var LEAVES_PER_NODE = 5;
+    var LEAVES_PER_PATH_LEVEL = LEAVES_PER_NODE - 1;
+
+    signal input leaf;
+    signal input path_index[levels];
+    signal input path_elements[levels][LEAVES_PER_PATH_LEVEL];
+    signal output root;
+
+    var i;
+    var j;
+
+    component hashers[levels];
+    component splicers[levels];
+
+    // Hash the first level of path_elements
+    splicers[0] = Splicer(LEAVES_PER_PATH_LEVEL);
+    hashers[0] = Hasher5();
+    splicers[0].index <== path_index[0];
+    splicers[0].leaf <== leaf;
+    for (i = 0; i < LEAVES_PER_PATH_LEVEL; i++) {
+        splicers[0].in[i] <== path_elements[0][i];
+    }
+
+    for (i = 0; i < LEAVES_PER_NODE; i++) {
+        hashers[0].in[i] <== splicers[0].out[i];
+    }
+
+    // Hash each level of path_elements
+
+    for (i = 1; i < levels; i++) {
+        splicers[i] = Splicer(LEAVES_PER_PATH_LEVEL);
+        splicers[i].index <== path_index[i];
+        splicers[i].leaf <== hashers[i - 1].hash;
+        for (j = 0; j < LEAVES_PER_PATH_LEVEL; j ++) {
+            splicers[i].in[j] <== path_elements[i][j];
+        }
+
+        hashers[i] = Hasher5();
+        for (j = 0; j < LEAVES_PER_NODE; j ++) {
+            hashers[i].in[j] <== splicers[i].out[j];
+        }
+    }
+    
+    root <== hashers[levels - 1].hash;
+}
+
+template QuinLeafExists(levels){
+    // Ensures that a leaf exists within a quintree with given `root`
+
+    var LEAVES_PER_NODE = 5;
+    var LEAVES_PER_PATH_LEVEL = LEAVES_PER_NODE - 1;
+
+    var i;
+    var j;
+
+    signal input leaf;
+    signal input path_elements[levels][LEAVES_PER_PATH_LEVEL];
+    signal input path_index[levels];
+    signal input root;
+
+    // Verify the Merkle path
+    component verifier = QuinTreeInclusionProof(levels);
+    verifier.leaf <== leaf;
+    for (i = 0; i < levels; i ++) {
+        verifier.path_index[i] <== path_index[i];
+        for (j = 0; j < LEAVES_PER_PATH_LEVEL; j ++) {
+            verifier.path_elements[i][j] <== path_elements[i][j];
+        }
+    }
+
+    root === verifier.root;
+}
+
+/*
+ * Given a tree index, generate the indices which QuinTreeInclusionProof and
+ * QuinLeafExists require. e.g. if the index is 30 and the number of levels is
+ * 4, the output should be [1, 1, 0, 0]
+ */
+template QuinGeneratePathIndices(levels) {
+    var BASE = 5;
+    signal input in; 
+    signal output out[levels];
+
+    var m = in;
+    for (var i = 0; i < levels; i ++) {
+        // circom's best practices state that we should avoid using <-- unless
+        // we know what we are doing. But this is the only way to perform the
+        // modulo operation. Later, we will do a range proof to ensure that the
+        // output is correct.
+        out[i] <-- m % BASE;
```

Comment:
@kobigurk @ChihChengLiang `<--` is usually a red flag for circom circuits, so I'm tagging you to seek your advice on whether the range proof below is sufficient :smile:

Reply:
LGTM!